math/fibonacci.c

```diff
@@ -1,23 +1,124 @@
+ * For further info: https://en.wikipedia.org/wiki/Fibonacci_sequence
+ *
+ * @author [Luiz Carlos Aguiar C](https://github.com/IKuuhakuI)
+ * @author [Niranjan](https://github.com/niranjank2022)
+ */
 
-// Fibonnacci function
-int fib(int number)
+#include <assert.h>  /// for assert()
+#include <errno.h>  /// for errno - to determine whether there is an using strtol()
+#include <stdio.h>   /// for input, output
+#include <stdlib.h>  /// for exit() - to exit the program
+#include <time.h>   /// to calculate time taken by fib()
+/**
+ * @brief Determines the nth Fibonacci term
+ * @param number - n in "nth term" and it can't be negative as well as zero
+ * @return nth term in unsigned type
+ * @warning
+ * Only till 47th and 48th fibonacci element can be stored in 'int' and
+ * 'unsigned' respectively (takes more than 20 seconds to print)
+ */
+unsigned int fib(int number)
 {
-    if (number == 1 || number == 2)
+    // Check for negative integers
+    if (number <= 0)
+    {
+        fprintf(stderr, "Illegal Argument Is Passed!\n");
+        exit(EXIT_FAILURE);
+    }
+
+    // Base conditions
+    if (number == 1)
+        return 0;
+
+    if (number == 2)
         return 1;
-    else
-        return fib(number - 1) + fib(number - 2);
+
+    // Recursive call to the function
+    return fib(number - 1) + fib(number - 2);
 }
 
+/**
+ * @brief Get the input from the user
+ * @return valid argument to the fibonacci function
+ */
+int getInput(void)
+{
+    int num, excess_len;
+    char buffer[3], *endPtr;
+
+    while (1)
+    {  // Repeat until a valid number is entered
+        printf("Please enter a valid number:");
+        fgets(buffer, 3, stdin);  // Inputs the value from user
+
+        excess_len = 0;
+        if (!(buffer[0] == '\n' ||
+            buffer[1] == '\n' ||
+            buffer[2] == '\n')) {
+            while (getchar() != '\n') excess_len++;
+        }
+
+        num = strtol(buffer, &endPtr,
+                     10);  // Attempts to convert the string to integer
+
+        // Checking the input
+        if (  // The number is too large
+            (excess_len > 0 || num > 48) ||
+            // Characters other than digits are included in the input
+            (*endPtr != '\0' && *endPtr != '\n') ||
+            // No characters are entered
+            endPtr == buffer)
+        {
+            continue;
+        }
+
+        break;
+    }
+
+    printf("\nEntered digit: %d (it might take sometime)\n", num);
+    return num;
+}
+
+/**
+ * @brief self-test implementation
+ * @return void
+ */
+static void test()
+{
+    assert(fib(5) == 3);
+    assert(fib(2) == 1);
+    assert(fib(9) == 21);
+}
+
+/**
+ * @brief Main function
+ * @return 0 on exit
+ */
 int main()
 {
-    int number;
+    // Performing the test
+    test();
+    printf("Tests passed...\n");
 
-    // Asks for the number that is in n position in Fibonnacci sequence
-    printf("Number: ");
-    scanf("%d", &number);
+    // Getting n
+    printf(
```

Comment:
We should probably check if the number's above 48, and if so, do not let the Fibonacci function run.

Reply:
Yes. And also not permit other invalid inputs

Reply:
I thought the `getinput()` function already took care of that? Or the idea is to stop the program instead of re-request input? I must say I'm a bit confused about what the two of you are talking about.

Reply:
Yeah, the function already manages that (kind of missed it out).